csharp_dotnetcore/13.Basic-Bot-Template/Dialogs/MainDialog/MainDialog.cs

```diff
@@ -0,0 +1,203 @@
+using Microsoft.BotBuilderSamples;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+
+namespace Microsoft.BotBuilderSamples
+{
+    public class MainDialog : ComponentDialog
+    {
+        // Supported LUIS Main Dialog Intents
+        public const string GreetingIntent = "Greeting";
+        public const string HelpIntent = "Help";
+
+        /// <summary>
+        /// Key in the bot config (.bot file) for the LUIS instance.
+        /// In the .bot file, multiple instances of LUIS can be configured.
+        /// </summary>
+        public static readonly string LuisKey = "BasicBotLUIS";
+
+        private readonly BotServices _services;
+        private readonly ILogger _logger;
+
+        public MainDialog(BotServices services, BasicBotAccessors accessors, ILogger logger)
+            : base(nameof(MainDialog))
+        {
+            _services = services ?? throw new ArgumentNullException(nameof(services));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+
+            AddDialog(new GreetingDialog(services, accessors.DialogStateProperty, accessors.GreetingStateProperty, logger));
+            AddDialog(new NamePrompt(nameof(NamePrompt)));
+            AddDialog(new CityPrompt(nameof(CityPrompt)));
+        }
+
+        protected override async Task<DialogTurnResult> OnDialogBeginAsync(DialogContext dc, DialogOptions options, CancellationToken cancellationToken)
+        {
+            // Override default begin() logic with bot orchestration logic
+            return await OnDialogContinueAsync(dc, cancellationToken).ConfigureAwait(false);
+        }
+
+        protected override async Task<DialogTurnResult> OnDialogContinueAsync(DialogContext dc, CancellationToken cancellationToken)
+        {
+            if (dc.Context.Activity.Type == ActivityTypes.Message)
+            {
+                // If an active dialog is waiting, continue dialog
+                var result = await dc.ContinueAsync();
+
+                switch (result.Status)
+                {
+                    case DialogTurnStatus.Empty:
+                        // No dialog is currently on the stack and we haven't responded to the user.
+                        // Perform a call to LUIS to retrieve results for the current activity message.
+                        var luisResults = await _services.LuisServices[LuisKey].RecognizeAsync(dc.Context, cancellationToken).ConfigureAwait(false);
+                        var topIntent = luisResults?.GetTopScoringIntent();
+                        var interruptResult = InterruptionStatus.NoAction;
+                        if (topIntent != null)
+                        {
+                            // See if there are any conversation interrupts we need to handle
+                            switch (topIntent.Value.intent)
+                            {
+                                case GreetingIntent:
+                                    if (topIntent.Value.score > .5)
```

Comment:
We should not show code that compares scores to constants. Actually we should only compare scores within a model.

In this particular case we are hitting up against the issue where LUIS is not going to work well for discriminating single word utterances. So for discriminating intent between, say, "cancel," "help" and "hi" we should not use LUIS instead we could use something like regular expressions or some other pattern match.